### PR TITLE
ABI hash based ids for Logs.

### PR DESCRIPTION
**Pull request by esdrubal**
## Description

Changes LogId to contain hash based on abi str.

```
    __log(1u32);         // hashed: "u32"
    __log((1u32, 1u64)); // hashed: "(u32, u64)"
    __log([0u64]);       // hashed: "[u64; 1]"
    __log(Some(0u64));   // hashed: "enum std::option::Option<u64>"
    __log(SS{
        ss: 1u64
    });                  // hashed: "struct SS<u64>"
```

Adds abi_hash_ids flag to command line tools.

Closes https://github.com/FuelLabs/sway/issues/5953

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
   - [ ] If my change requires substantial documentation changes, I have [requested support from the DevRel team](https://github.com/FuelLabs/devrel-requests/issues/new/choose)
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: test/src/e2e_vm_tests/test_programs/should_pass/language/logging/json_abi_oracle.json [hunk 1]
```diff
@@ -1,79 +1,56 @@
...
+      "logId": 5087777005172090899,
       "loggedType": {
         "name": "",
         "type": 1,
         "typeArguments": []
       }
     },
     {
-      "logId": 4,
+      "logId": 5555909392781521367,
```
Comment (Voxelot) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/logging/json_abi_oracle.json:48]: hmmm, technically js only support 52 bit numbers in json. So we either need to make this a string (which is breaking), or we have to truncate this hash further.

Comment (arboleya) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/logging/json_abi_oracle.json:48]: Good catch! Making it a `string` should be easy enough.

cc @nedsalk

File: forc-test/src/lib.rs [hunk 2]
```diff
@@ -683,11 +683,11 @@ pub fn decode_log_data(
         .logged_types
         .iter()
         .flatten()
-        .map(|logged_type| (logged_type.log_id.as_str(), logged_type.application.clone()))
+        .map(|logged_type| (logged_type.log_id.clone(), logged_type.application.clone()))
```
Comment (kayagokalp) [inline forc-test/src/lib.rs:686]: while rebasing accepted @esdrubal's changes blindly. I am reverting this one to mine, as it does not clone

File: test/src/e2e_vm_tests/test_programs/should_pass/language/logging/json_abi_oracle.json [hunk 3]
```diff
@@ -1,79 +1,56 @@
...
-      "logId": 2,
-      "loggedType": {
-        "name": "",
-        "type": 1,
-        "typeArguments": []
-      }
-    },
-    {
-      "logId": 3,
+      "logId": 5087777005172090899,
```
Comment (Voxelot) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/logging/json_abi_oracle.json:40]: @kayagokalp not sure why this didn't get caught by tests, but these are still non-str values

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/assert_eq_revert/json_abi_oracle.json [hunk 4]
```diff
@@ -14,15 +15,7 @@
...
-      "logId": 0,
-      "loggedType": {
-        "name": "",
-        "type": 1,
-        "typeArguments": null
-      }
-    },
-    {
-      "logId": 1,
+      "logId": 1515152261580153489,
```
Comment (Voxelot) [inline test/src/e2e_vm_tests/test_programs/should_pass/stdlib/assert_eq_revert/json_abi_oracle.json:18]: also still using integer values

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/assert_ne/json_abi_oracle.json [hunk 5]
```diff
@@ -14,175 +15,79 @@
   ],
   "loggedTypes": [
     {
-      "logId": 0,
+      "logId": 1515152261580153489,
```
Comment (Voxelot) [inline test/src/e2e_vm_tests/test_programs/should_pass/stdlib/assert_ne/json_abi_oracle.json:18]: still using integers

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/assert_ne_revert/json_abi_oracle.json [hunk 6]
```diff
@@ -14,15 +15,7 @@
...
-      "logId": 0,
-      "loggedType": {
-        "name": "",
-        "type": 1,
-        "typeArguments": null
-      }
-    },
-    {
-      "logId": 1,
+      "logId": 1515152261580153489,
```
Comment (Voxelot) [inline test/src/e2e_vm_tests/test_programs/should_pass/stdlib/assert_ne_revert/json_abi_oracle.json:18]: and here

File: test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/basic_storage/json_abi_oracle.json [hunk 7]
```diff
@@ -198,95 +199,23 @@
   ],
   "loggedTypes": [
     {
-      "logId": 0,
+      "logId": 1515152261580153489,
```
Comment (Voxelot) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/basic_storage/json_abi_oracle.json:202]: and here

File: test/src/e2e_vm_tests/test_programs/should_pass/language/logging/json_abi_oracle.json (hunk 3 above)
Comment (kayagokalp) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/logging/json_abi_oracle.json:40]: That is interesting, investigating now. Maybe during rebasing I mixed up things and modified the oracle invalidly somehow. But the fact that `output` file is generated as using strings tells me the code itself is working correctly. 

It is weird that this is not caught by the test suite.

Comment (kayagokalp) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/logging/json_abi_oracle.json]: Not sure how this is not caught. it seems like test suite is doing string comparison between oracle and the output

https://github.com/FuelLabs/sway/blob/2ec4332c550e3bad4cc795b189342f1d52b14ebf/test/src/e2e_vm_tests/harness.rs#L398

**Review by IGI-111 — APPROVED** [review]

**Review by nedsalk — APPROVED** [review]
I tried out the branch locally with the TS-SDK and it worked. Approving as-is after chat with @IGI-111.

**Pull request merged**